### PR TITLE
Handle virtual beatmap tracks more gracefully

### LINKED ISSUE #2770: Beatmaps without Mp3 causing a crash
Beatmaps with no Mp3 causing lazer to crash when you try to change the state in the music controller 
[runtime.log](https://github.com/ppy/osu/files/2087555/runtime.log)

### PR DESCRIPTION
**Pull request by smoogipoo**
- Fixes #2770

---

Some beatmaps that don't have tracks could cause osu! to crash due to an invalid audio length. This change calculates the virtual track length based on the last hitobject's time, circumventing any possible crashes.

### CODE REVIEW

**Comment by Aergwyn** [comment]
This could potentially fix two major issues if you have no audio device.
a) The `MusicController` going out of control -> it spams `Next` because all songs had 0 length
b) Not being able to start a map at all -> pressing play gives the `MusicController` time to trigger `Next` once (see a) which either starts a song with a wrong map or crashes

File: osu.Game/Beatmaps/WorkingBeatmap_VirtualBeatmapTrack.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+            public VirtualBeatmapTrack(IBeatmap beatmap)
+            {
+                this.beatmap = beatmap;
+                updateVirtualLength();
+            }
+
+            protected override void UpdateState()
+            {
+                updateVirtualLength();
```
Comment (peppy) [inline osu.Game/Beatmaps/WorkingBeatmap_VirtualBeatmapTrack.cs]: why is the update here required? shouldn't `beatmap` never change?

Comment (smoogipoo) [reply osu.Game/Beatmaps/WorkingBeatmap_VirtualBeatmapTrack.cs]: Editor?

Comment (peppy) [reply osu.Game/Beatmaps/WorkingBeatmap_VirtualBeatmapTrack.cs]: i'd wait until we have the editor actually performing operations. we likely don't want this kind of thing running per-frame (but on a callback instead).

File: osu.Game/Beatmaps/WorkingBeatmap_VirtualBeatmapTrack.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            }
+
+            private void updateVirtualLength()
+            {
+                var lastObject = beatmap.HitObjects.LastOrDefault();
+
+                switch (lastObject)
+                {
+                    case null:
+                        Length = 1000;
```
Comment (peppy) [inline osu.Game/Beatmaps/WorkingBeatmap_VirtualBeatmapTrack.cs]: `const double excess_length = 1000;`

**Comment by smoogipoo** [comment]
CI doesn't have an audio device, so the enqueued track actions never run. Technically the second instance in that commit doesn't need it since it doesn't invoke a method, but the seek() does.

The commit preserves previous functionality.

**Review by peppy — APPROVED** [review]

**Pull request merged**